src/gt4py/next/program_processors/runners/dace_iterator/itir_to_tasklet.py

```diff
@@ -350,6 +351,104 @@ def builtin_undefined(*args: Any) -> Any:
 }
 
 
+class GatherLambdaSymbolsPass(eve.NodeVisitor):
+    sdfg: dace.SDFG
+    state: dace.SDFGState
+    symbol_map: dict[str, IteratorExpr | SymbolExpr | ValueExpr]
+    parent_symbol_map: dict[str, IteratorExpr | SymbolExpr | ValueExpr]
+
+    def __init__(
+        self,
+        sdfg,
+        state,
+        symbol_map,
+        parent_symbol_map,
+    ):
+        self.sdfg = sdfg
+        self.state = state
+        self.symbol_map = symbol_map
+        self.parent_symbol_map = parent_symbol_map
+
+    def _add_symbol(self, param, arg):
+        if isinstance(arg, ValueExpr):
+            # create storage in lambda sdfg
+            self.sdfg.add_scalar(param, dtype=arg.dtype)
+            # update table of lambda symbol
+            self.symbol_map[param] = ValueExpr(self.state.add_access(param), arg.dtype)
+        elif isinstance(arg, IteratorExpr):
+            # create storage in lambda sdfg
+            ndims = len(arg.dimensions)
+            shape = tuple(
+                dace.symbol(unique_var_name() + "__shp", dace.int64) for _ in range(ndims)
+            )
+            strides = tuple(
+                dace.symbol(unique_var_name() + "__strd", dace.int64) for _ in range(ndims)
+            )
+            self.sdfg.add_array(param, shape=shape, strides=strides, dtype=arg.dtype)
+            index_names = {dim: f"__{param}_i_{dim}" for dim in arg.indices.keys()}
+            for _, index_name in index_names.items():
+                self.sdfg.add_scalar(index_name, dtype=dace.int64)
+            # update table of lambda symbol
+            field = self.state.add_access(param)
+            indices = {
+                dim: self.state.add_access(index_arg) for dim, index_arg in index_names.items()
+            }
+            self.symbol_map[param] = IteratorExpr(field, indices, arg.dtype, arg.dimensions)
+        else:
+            assert isinstance(arg, SymbolExpr)
+            self.symbol_map[param] = arg
+
+    def visit_SymRef(self, node: itir.SymRef):
+        name = str(node.id)
+        if name in self.parent_symbol_map and name not in self.symbol_map:
+            arg = self.parent_symbol_map[name]
+            self._add_symbol(name, arg)
+
+    def visit_FunCall(self, node: itir.FunCall):
+        self.visit(node.args)
+        self.visit(node.fun)
+
+    def visit_Lambda(self, node: itir.Lambda, args=None):
```

Comment:
Could you add type annotations for `args`?

I'm also thinking if we should make this parameter mandatory and pass an empty list/dict/whatever it is.

Reply:
I have added the type annotation for now. I will investigate the other proposal.

Reply:
We need to keep `args` defaulted to `None`. We could enter this function either from the root lambda node, for which we have argument nodes, or from a nested lambda, for which we do not want to pass the arguments because the argument nodes for a nested lambda should not appear in the symbol map of the root lambda.

Reply:
Okay, I see